### PR TITLE
refactor(open_graph): drop cheerio and use regex

### PR DESCRIPTION
**Pull request by SukkaW**
## What does it do?

cheerio is slow and we only use that to parse and list img in post.content. This PR drop cheerio and use regex to do the job.

## How to test

```sh
git clone -b BRANCH https://github.com/USER/hexo.git
cd hexo
npm install
npm test
```

## Screenshots
 
https://github.com/hexojs/hexo/pull/3680/commits/913f5bed0d39cacda66479bd58345992b64cfb8d

| | NodeJS 8 | NodeJS 10 | NodeJS 12 |
|:---:|:---:|:---:|:---:|
| cheerio@0.22 (hexojs/hexo#master) | **7.12** | 6.96 | 6.86 |
| regex (This PR) | 7.15 | 6.96 | **6.78** |
| dom-parser (#3679) | 7.2 | 6.93 | 6.82 |
| conditional cheerio (#3670) | 7.15 | **6.92** | **6.78** |

https://github.com/hexojs/hexo/pull/3680/commits/4f86ddef8f8f70193c4b590ac35d0eaad9028237

| | NodeJS 8 | NodeJS 10 | NodeJS 12 |
|:---:|:---:|:---:|:---:|
| cheerio@0.22 (hexojs/hexo#master) | 7.03 | 6.59 | 6.51 |
| regex (This PR) | 7.01 | 6.57 | 6.45 |
| dom-parser (#3679) | **6.98** | **6.56** | 6.41 |
| conditional cheerio (#3670) | 7.00 | 6.60 | **6.38** |

https://github.com/hexojs/hexo/pull/3680/commits/5c309eb109c55f78fd5c4c27f070e932c9150b03

| | NodeJS 8 | NodeJS 10 | NodeJS 12 |
|:---:|:---:|:---:|:---:|
| cheerio@0.22 (hexojs/hexo#master) | 6.96 | 6.65 | 6.85 |
| regex (This PR) | 6.90 | 6.68 | **6.73** |
| dom-parser (#3679) | 6.88 | 6.67 | **6.73** |
| conditional cheerio (#3670) | **6.72** | **6.60** | 6.80 |

https://github.com/hexojs/hexo/pull/3680/commits/c11ded4735d702890cc0f42298fcf1cde53b4c15

| | NodeJS 8 | NodeJS 10 | NodeJS 12 |
|:---:|:---:|:---:|:---:|
| cheerio@0.22 (hexojs/hexo#master) | 7.29 | 6.54 | 6.93 |
| regex (This PR) | 7.21 | **6.47** | 6.84 |
| dom-parser (#3679) | 7.16 | 6.50 | 6.87 |
| conditional cheerio (#3670) | **7.06** | 6.50 | **6.81** |

## Pull request tasks

- [ ] Add test cases for the changes.
- [x] Passed the CI test.

### CODE REVIEW

**Comment by SukkaW** [comment]
For regex, I have test those pattern with [regex101](https://regex101.com/) to match [this post](https://blog.skk.moe/post/how-verizon-knocked-large-parts-of-the-internet-offline/):

```
/<img.*?src="(.*?)".*?\/?>/gi
```

6491 steps, 9 matches, ~3ms: https://regex101.com/r/N1Rv9w/1

```
/<img.*?(?:>|\/>)/gi
```

8858 steps, 9 matches, ~5ms: https://regex101.com/r/N1Rv9w/2

```
/<img(.*?)>/gi
```

6527 steps, 9 matches, ~4ms: https://regex101.com/r/N1Rv9w/3

```
<img.*?>
```

4313 steps, 9 matched, ~3ms: https://regex101.com/r/N1Rv9w/4

```
<img[^>]*src[="'\s]+[^.]*\/([^.]+)\.[^"']+["']?[^>]*>
```

4627 steps, 9 matched, ~4ms: https://regex101.com/r/N1Rv9w/5

Currently in this PR I am using the first one.

File: lib/plugins/helper/open_graph.js [hunk 1]
```diff
@@ -59,12 +57,18 @@ function openGraphHelper(options = {}) {
...
     images = images.slice();
 
-    const $ = cheerio.load(content);
+    if (content.includes('<img')) {
+      const imgPattern = /<img.*?>/gi;
+      const srcPattern = /src=['"]?([^'"]*)['"]?/i;
+      const imgArray = content.match(imgPattern);
+      for (let img of imgArray) {
+        const src = img.match(srcPattern);
+        if (src[1]) {
```
Comment (segayuu) [inline lib/plugins/helper/open_graph.js]: Since the img element requires the src attribute, the null check of String # match () will not be necessary in the normal system.
If there is an img element without src, it should be treated as abnormal, but there is room for consideration in error handling.

Comment (SukkaW) [reply lib/plugins/helper/open_graph.js]: Since it is using regex rather than a dom parser to match the `` tag, so an `` tag without src could happened. If there is an `` tag matches `imgPattern` but without `src` attribute, `img.match(srcPattern)` will return `null`. So should I use `if (src)` rather than `if (src[1])`?

Comment (segayuu) [reply lib/plugins/helper/open_graph.js]: If you want to ignore an `img` element with an empty `src`, the following changes will eliminate unnecessary checks.
`const srcPattern = /src=['"]?([^'"]+)['"]?/i;`
`if (src) images.push(src[1]); `

Comment (SukkaW) [reply lib/plugins/helper/open_graph.js]: OK, I will change to that.

Comment (segayuu) [reply lib/plugins/helper/open_graph.js]: Thank you for the correction.
However, with just this one line, even if the value of `src` is empty, it will be pushed to `images`.
If the value of `src` is intended to be ignored if it is empty, you need to change the definition of the `srcPattern` variable.

**Comment by SukkaW** [comment]
I think we still need to dig into those PR: #3679 #3670 and this PR. They all increase performance, but not much.

File: lib/plugins/helper/open_graph.js (hunk 1 above)
Comment (SukkaW) [reply lib/plugins/helper/open_graph.js]: I have pushed a new commit with a new regex pattern:

```
/<img [^>]*src=['"]([^'"]+)([^>]*>)/gi
```

This one will only match the `` tag with non-empty `src` attribute.

**Comment by curbengh** [comment]
Since this PR only uses native API, naturally it supersedes #3679 #3670. This is unless there is possibility of some edge cases, which I currently don't foresee.

**Comment by SukkaW** [comment]
@segayuu @curbengh Should I add some test cases for `` tags with empty `src` attribute?

**Comment by SukkaW** [comment]
I have update the final benchmark for this PR & #3670

| | NodeJS 8 | NodeJS 10 | NodeJS 12 |
|:---:|:---:|:---:|:---:|
| cheerio@0.22 (hexojs/hexo#master) | 8.75 | 7.0 | 7.82 |
| regex (This PR) | **6.51** | **5.52** | **6.44** |
| conditional cheerio (#3670) | 7.68 | 6.49 | 7.51 |

The last benchmark shows this PR increase performance by nearly `20%`.

I will merge this PR and close #3670 then.
